make performancetips code consistent with offset arrays

mkborregaard opened the pull request:
As remarked on Discourse (https://discourse.julialang.org/t/is-inbounds-really-necessary/5654/12) the performance tips code for inbounds is not consistent with recommended style.

The performance tips e.g. use 1:n to iterate over an array to demonstrate a case where it is safe to use `@inbounds`. But that very example is used here: https://docs.julialang.org/en/latest/devdocs/offset-arrays/#Background-1 to describe a situation where it is NOT safe to use `@inbounds` because the array could be an offsetarray.

When used as in the context here it is still safe, because the array is initialized in the previous line, but it is possibly misleading.

I updated the code examples on the page to be consistent with the generalized-array syntax as I understand it.

mkborregaard committed:
make consistent with offset arrays

timholy commented:
Thanks! LGTM.

Anything that specifies `Vector` or `Array` in its argument types will predictably start indexing at 1, nevertheless :+1: for demonstrating more generalizable coding practices.

andreasnoack commented on doc/src/manual/performance-tips.md:
```diff
@@ -805,36 +805,36 @@ do this in at least four ways (in addition to the recommended call to the built-
-    n = size(x, 1)
-    out = Array{T}(n, n)
-    for i = 1:n
+    n = indices(x, 1)
```
Maybe change the name of the variable to e.g. `inds`.

mkborregaard replied:
Yes, good point. I made the change.

mkborregaard committed:
replace `n` with `inds`

mkborregaard commented on doc/src/manual/performance-tips.md:
```diff
@@ -1174,31 +1174,31 @@ of a one-dimensional array, and then evaluates the L2-norm of the result:
-    @fastmath @inbounds @simd for i in 2:n-1
+    inds = linearindices(u)
+    dx = 1.0 / (length(inds)-1)
+    @fastmath @inbounds du[inds[1]] = (u[inds[2]] - u[inds[1]]) / dx
```
In most cases the more general style is at least as clear - but in this one case it isn't, and I'm unsure whether this might confuse readers? I tried to be as clear as possible.

KristofferC commented:
I think this might just detract from the point of the examples. I rather just have a `::Array` annotation on the input variable or a statement saying that we assume `Arrays` in this section and a link to the general indices section. Thoughts?

mkborregaard commented:
I think that could be true for the line I highlighted above - that's why I highlighted it. I was thinking to suggest annotating that with ::Array. In the other examples I think the style is at least as clear and might be said to be a preferred style anyway, at least it does not detract from clarity IMHO.

I'd say it is an issue that the very first example for inbounds here https://docs.julialang.org/en/latest/manual/performance-tips/#man-performance-annotations-1 is exactly a case that this part https://docs.julialang.org/en/latest/devdocs/offset-arrays/#Background-1 highlights as unsafe.

In the end I guess it depends how much you'd like the generalized array interface to permeate the language and ecosystem. We're now editing Plots to conform with the generalized syntax and it does require a slightly different thinking.

timholy commented:
In places where it's awkward, I'm fine with restricting it to Arrays.

But overall I agree that we should demo code that uses the generic interface. When I get some time again I plan to "take off the training wheels" and submit a PR to OffsetArrays that allows `@inbounds` to work and supports `size` and `length` again (and submit a PR to the devdocs updating the recommendations). At that point one is back in segfault-land for code that hasn't been updated.

mkborregaard committed:
Replace some instances with Array assertion

Replaces generalized indexing syntax with an ::Vector assertion where that improved clarity of the code.

mkborregaard committed:
Link to generalized indexing page

mkborregaard commented:
I've 1. inserted an ::Vector assertion in the place where the code was awkward, and 2. mentioned the concern with unconventional indexing when using inbounds. This should meet the concerns.

JeffBezanson commented on doc/src/manual/performance-tips.md:
```diff
@@ -1094,6 +1094,10 @@ Sometimes you can enable better optimization by promising certain program proper
   * Write `@simd` in front of `for` loops that are amenable to vectorization. **This feature is experimental**
     and could change or disappear in future versions of Julia.
 
+The common ideom of using 1:n to index into an AbstractArray is not safe if the Array uses unconventional indexing,
```
idiom

mkborregaard replied:
Fixed.

JeffBezanson commented on doc/src/manual/performance-tips.md:
````diff
@@ -601,7 +601,7 @@ This should be written as:
 ```jldoctest
 julia> function fill_twos!(a)
-           for i=1:length(a)
+           for i=linearindices(a)
````
Would it make more sense to use `eachindex(a)`?

mkborregaard replied:
agreed that `eachindex` is more intuitive. But the offset-arrays doc recommend `linearindices` (https://docs.julialang.org/en/latest/devdocs/offset-arrays) so I chose that for consistency.
Should I
1. keep `linearindices` both places for consistency
2. change `linearindices`->`eachindex` both places for consistency (for this use - iterating linearly over a container)
3. change `linearindices`->`eachindex` here and not worry about the small discrepancy
?
I have no clear preference.

timholy replied:
Use `linearindices` if you really want to index with an integer. If you don't care what you index with, use `eachindex`.

timholy replied:
I see what you mean about it being ambiguous in the devdocs.

mkborregaard replied:
I've updated to `eachindex` here. I don't mind including some changes to the devdocs here, or we sould call it good as it is?

timholy replied:
If you're willing I'd be grateful, but feel free to say "your job, Tim." :smile:

mkborregaard replied:
I've tried here: https://github.com/JuliaLang/julia/pull/23731 . Can't say I'm completely happy with the result at this point, but possibly after some feedback that could work.

mkborregaard committed:
fix typo ideom -> idiom

mkborregaard committed:
replace linearindices with eachindex

mbauman committed:
Remove trailing whitespace

timholy commented:
Thanks so much, @mkborregaard!
